src/actions/marketo/queue.ts

```diff
@@ -0,0 +1,135 @@
+// function log(...args: any[]) {
+//   console.log(...args)
+// }
+
+// function logJson(label: string, object: any) {
+//   console.log("\n================================")
+//   console.log(`${label}:\n`)
+//   const json = `${JSON.stringify(object)}\n\n`
+//   console.log(json)
+// }
+
+interface Task {
+  id: number
+  run: any
+  result?: any
+  error?: any
+}
+
+/**
+ * Queue
+ * helper to run 10 tasks at a time
+ *
+ * Usage:
+ *
+ * const queue = new Queue()
+ *
+ * queue.addTask(async fn)
+ * queue.addTask(async fn)
+ * queue.addTask(async fn)
+ * queue.addTask(async fn)
+ * ... add as many tasks as needed
+ *
+ * const completed = await queue.finish()
+ *
+ * `completed` is an array of objects with { result } or { error }
+ * (the resolved result of each task or the rejected result)
+ * the completed results will be in the same order as they were added
+ *
+ */
+export class Queue {
+
+  channelSize: number
+  counter: number
+  queue: Task[]
+  channels: Task[]
+  completed: Task[]
+  finished: boolean
+  promise: Promise<Task[]>
+  resolve: any
+
+  constructor() {
+    this.channelSize = 10 // TODO make this configurable?
```

Comment:
10 concurrent requests per subscription looks good to me per Marketos docs. One thing that concerns me though is the API rate-limiting. Have you tested a large enough dataset to confirm that we do not send more than 100 API calls per 20 seconds? http://developers.marketo.com/rest-api/marketo-integration-best-practices/

Reply:
Hey Phillipe, yes our test data set was large enough to test the rate-limiting with no issues, although there's nothing in the code to prevent our hitting the limit if for some reason Marketo's API decides to get super responsive.